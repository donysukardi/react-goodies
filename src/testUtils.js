const propsToDataAttrs = props =>
  Object.entries(props).reduce((acc, [key, value]) => {
    acc[`data-${String(key).toLowerCase()}`] = value
    return acc
  }, {})

export {propsToDataAttrs}
