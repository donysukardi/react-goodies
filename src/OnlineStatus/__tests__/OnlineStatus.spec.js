/* eslint-disable global-require */
import React from 'react'
import { cleanup, render } from 'react-testing-library'
import { propsToDataAttrs } from '../../testUtils'

const Target = props => (
  <div data-testid="targetContainer" {...propsToDataAttrs(props)} />
)

const renderTarget = props => <Target {...props} />

describe('OnlineStatus', () => {
  describe('Online Status API is supported', () => {
    let origOnlineStatus = null
    let OnlineStatus = null

    afterEach(cleanup)

    beforeAll(() => {
      origOnlineStatus = global.navigator.onLine
    })

    beforeEach(() => {
      jest.resetModules()

      Object.defineProperty(global.navigator, 'onLine', {
        get: () => true,
        configurable: true,
      })
      OnlineStatus = require('../OnlineStatus').default
    })

    afterAll(() => {
      Object.defineProperty(global.navigator, 'onLine', {
        get: () => origOnlineStatus,
        configurable: true,
      })
    })

    it('should set initial state', () => {
      const { getByTestId } = render(
        <OnlineStatus>{renderTarget}</OnlineStatus>,
      )

      expect(getByTestId('targetContainer')).toMatchSnapshot()
    })

    it('should handle online status change', () => {
      const { getByTestId } = render(
        <OnlineStatus>{renderTarget}</OnlineStatus>,
      )

      global.dispatchEvent(new CustomEvent('offline'))
      expect(getByTestId('targetContainer')).toMatchSnapshot()
      global.dispatchEvent(new CustomEvent('online'))
      expect(getByTestId('targetContainer')).toMatchSnapshot()
    })

    it('should remove event listener on unrender', () => {
      const spy = jest.spyOn(global, 'removeEventListener')

      const { unmount } = render(<OnlineStatus>{renderTarget}</OnlineStatus>)
      unmount()

      global.document.dispatchEvent(new CustomEvent('offline'))

      expect(spy.mock.calls).toMatchSnapshot()

      global.removeEventListener.mockRestore()
    })
  })
})
