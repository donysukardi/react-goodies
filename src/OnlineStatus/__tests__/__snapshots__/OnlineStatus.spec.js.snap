// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OnlineStatus Online Status API is supported should handle online status change 1`] = `
<div
  data-isoffline="true"
  data-isonline="false"
  data-testid="targetContainer"
/>
`;

exports[`OnlineStatus Online Status API is supported should handle online status change 2`] = `
<div
  data-isoffline="false"
  data-isonline="true"
  data-testid="targetContainer"
/>
`;

exports[`OnlineStatus Online Status API is supported should remove event listener on unrender 1`] = `
Array [
  Array [
    "error",
    [Function],
  ],
  Array [
    "error",
    [Function],
  ],
  Array [
    "error",
    [Function],
  ],
  Array [
    "error",
    [Function],
  ],
  Array [
    "online",
    [Function],
  ],
  Array [
    "offline",
    [Function],
  ],
  Array [
    "error",
    [Function],
  ],
  Array [
    "error",
    [Function],
  ],
  Array [
    "error",
    [Function],
  ],
]
`;

exports[`OnlineStatus Online Status API is supported should set initial state 1`] = `
<div
  data-isoffline="false"
  data-isonline="true"
  data-testid="targetContainer"
/>
`;
